x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/RollupDataExtractorFactory.java

```diff
@@ -0,0 +1,219 @@
+                              Job job,
+                              Map<String, RollableIndexCaps> rollupJobsWithCaps,
+                              ActionListener<DataExtractorFactory> listener) {
+
+        final AggregationBuilder datafeedHistogramAggregation = getHistogramAggregation(
+            datafeed.getAggregations().getAggregatorFactories());
+        if ((datafeedHistogramAggregation instanceof DateHistogramAggregationBuilder) == false) {
+            listener.onFailure(
+                new IllegalArgumentException("Rollup requires that the datafeed configuration use a [date_histogram] aggregation," +
+                    " not a [histogram] aggregation over the time field."));
+            return;
+        }
+
+        final String timeField = ((ValuesSourceAggregationBuilder) datafeedHistogramAggregation).field();
+
+        Set<ParsedRollupCaps> rollupCapsSet = rollupJobsWithCaps.values()
+            .stream()
+            .flatMap(rollableIndexCaps -> rollableIndexCaps.getJobCaps().stream())
+            .map(rollupJobCaps -> ParsedRollupCaps.fromJobFieldCaps(rollupJobCaps.getFieldCaps(), timeField))
+            .collect(Collectors.toSet());
+
+        final long datafeedInterval = getHistogramIntervalMillis(datafeedHistogramAggregation);
+
+        List<ParsedRollupCaps> validIntervalCaps = rollupCapsSet.stream()
+            .filter(rollupCaps -> validInterval(datafeedInterval, rollupCaps))
+            .collect(Collectors.toList());
+
+        if (validIntervalCaps.isEmpty()) {
+            listener.onFailure(
+                new IllegalArgumentException(
+                    "Rollup capabilities do not have a [date_histogram] aggregation with an interval " +
+                        "that is a multiple of the datafeed's interval.")
+            );
+            return;
+        }
+        final List<AggregationBuilder> flattenedAggs = new ArrayList<>();
+        flattenAggregations(datafeed.getAggregations().getAggregatorFactories(), datafeedHistogramAggregation, flattenedAggs);
+
+        if (validIntervalCaps.stream().noneMatch(rollupJobConfig -> hasAggregations(rollupJobConfig, flattenedAggs))) {
+            listener.onFailure(
+                new IllegalArgumentException("Rollup capabilities do not support all the datafeed aggregations at the desired interval.")
+            );
+            return;
+        }
+
+        listener.onResponse(new RollupDataExtractorFactory(client, datafeed, job));
+    }
+
+    private static boolean validInterval(long datafeedInterval, ParsedRollupCaps rollupJobGroupConfig) {
+        if (rollupJobGroupConfig.hasDatehistogram() == false) {
+            return false;
+        }
+        if (DateTimeZone.UTC.toString().equalsIgnoreCase(rollupJobGroupConfig.getTimezone()) == false) {
+            return false;
+        }
+        try {
+            long jobInterval = validateAndGetCalendarInterval(rollupJobGroupConfig.getInterval());
```

Comment:
nit: I find that when using static methods like in this case, it makes the code more readable to explicitly call `ExtractorUtils.validateAndGetCalendarInterval(...)` rather than statically import the method. It means I don't have to wonder where this method comes from. I suppose if a method was used many times in a file it could be otherwise. Anyway, not something you need to change, just a thought.